vm-virtio/src/vhost_user/blk.rs

```diff
@@ -0,0 +1,332 @@
+// SPDX-License-Identifier: Apache-2.0
+use libc;
+use libc::EFD_NONBLOCK;
+use std::cmp;
+use std::io::Write;
+use std::os::unix::io::AsRawFd;
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+use std::thread;
+use std::vec::Vec;
+
+use crate::VirtioInterrupt;
+
+use vm_memory::{Address, Error as MmapError, GuestMemory, GuestMemoryMmap, GuestMemoryRegion};
+use vmm_sys_util::eventfd::EventFd;
+
+use super::super::{ActivateError, ActivateResult, Queue, VirtioDevice, VirtioDeviceType};
+use super::handler::VhostUserEpollHandler;
+use super::{Error, Result};
+use vhost_rs::vhost_user::Master;
+use vhost_rs::vhost_user::VhostUserMaster;
+use vhost_rs::vhost_user::message::VhostUserConfigFlags;
+use vhost_rs::{VhostBackend, VhostUserMemoryRegionInfo, VringConfigData};
+use vhost_rs::vhost_user::message::{VhostUserProtocolFeatures, VhostUserVirtioFeatures};
+use virtio_bindings::virtio_blk;
+use std::mem;
+
+pub const VIRTIO_F_VERSION_1_BITMASK: u64 = 1 << VIRTIO_F_VERSION_1;
+pub const VIRTIO_F_VERSION_1: ::std::os::raw::c_uint = 32;
+
+macro_rules! offset_of {
+    ($ty:ty, $field:ident) => {
+        unsafe { &(*(0 as *const $ty)).$field as *const _ as usize }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+#[repr(C, packed)]
+struct virtio_blk_config {
+    capacity: u64,
+    size_max: u32,
+    seg_max: u32,
+    cylinders: u16,
+    heads: u8,
+    sectors: u8,
+    blk_size: u32,
+    physical_block_exp: u8,
+    alignment_offset: u8,
+    min_io_size: u16,
+    opt_io_size: u32,
+    wce: u8,
+    unused0: [u8; 1],
+    num_queues: u16,
+    max_discard_sectors: u32,
+    max_discard_seg: u32,
+    discard_sector_alignment: u32,
+    max_write_zeroes_sectors: u32,
+    max_write_zeroes_seg: u32,
+    write_zeroes_may_unmap: u8,
+    unused1: [u8; 3],
+}
+
+pub struct Blk {
+    vhost_user_blk: Master,
+    kill_evt: EventFd,
+    avail_features: u64,
+    acked_features: u64,
+    config_space: Vec<u8>,
+    queue_sizes: Vec<u16>,
+}
+
+impl Blk {
+    /// Create a new vhost-user-blk device
+    pub fn new(path: &str, num_queues: usize, queue_size: u16, config_wce: u8) -> Result<Blk> {
+        // Connect to the vhost-user socket.
+        let mut vhost_user_blk =
+            Master::connect(path, num_queues as u64).map_err(Error::VhostUserCreateMaster)?;
+
+        let kill_evt = EventFd::new(EFD_NONBLOCK).map_err(Error::CreateKillEventFd)?;
+        // Retrieve available features only when connecting the first time.
+        let mut avail_features = vhost_user_blk.get_features().map_err(Error::VhostUserGetFeatures)?;
+        // Let only ack features we expect, that is VIRTIO_F_VERSION_1.
+        if (avail_features & VIRTIO_F_VERSION_1_BITMASK) != VIRTIO_F_VERSION_1_BITMASK {
+            return Err(Error::InvalidFeatures);
+        }
+        avail_features =
+            VIRTIO_F_VERSION_1_BITMASK | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
+        vhost_user_blk
+            .set_features(avail_features)
```

Comment:
As mentioned on #69, I'd like to see this features negotiation being factorized so that both vu_fs, vu_net and vu_blk can use it.